File: sdk/java-sdk/src/main/java/com/evernym/verity/sdk/utils/DbcUtil.java
package com.evernym.verity.sdk.utils;

/**
 * Utilities for design by contract. Allows for common pre-condition and post-condition checking.
 */
public class DbcUtil {
    protected DbcUtil() {}

    /**
     * Checks and throws an IllegalArgumentException if the given object is null
     * @param arg any object that can be null
     */
    public static void requireNotNull(Object arg) {
        requireNotNull(arg, "ARG");
    }

    /**
     * Checks and throws an IllegalArgumentException if the given object is null
     * @param arg any object that can be null
     * @param argName name of the argument being checked, used in exception message if null
     */
    public static void requireNotNull(Object arg, String argName) {
        require(arg != null, "required that "+ argName +" must NOT be null");
    }

    /**
     * Checks that given boolean is true, throws an IllegalArgumentException if false
     * @param requirement testable requirement
     */
    public static void require(Boolean requirement) {
        require(requirement, "");
    }

    /**
     * Checks that given boolean is true, throws an IllegalArgumentException if false
     * @param requirement testable requirement
     * @param msg string that is used to build exception message
     */
    public static void require(Boolean requirement, String msg) {
        if(!requirement) {
            throw new IllegalArgumentException("requirement failed: "+ msg);
        }
    }

    /**
     * Checks and throws an IllegalArgumentException if the given string is null or empty
     * @param arg String argument which needs to be checked.
     * @param argName name of the argument being checked, used in exception message if null or empty
     */
    public static void requireStringNotNullOrEmpty(String arg, String argName) {
        require(arg != null && !arg.isEmpty(), "required that "+ argName +" must NOT be null or empty");
    }

    /**
     * Checks and throws an IllegalArgumentException if the given array is null or contains null element
     * @param array Array argument which needs to be checked.
     * @param argName name of the argument being checked, used in exception message
     */
    public static void requireArrayNotContainNull(Object[] array, String argName) {
        requireNotNull(array, argName);
        for (Object x: array)
            require(x != null, "required that elements of "+ argName +" must NOT be null");
    }
}

File: sdk/python-sdk/verity_sdk/utils/DbcUtil.py
class DbcUtil:
    """
    Utilities
    for design by contract.Allows for common pre-condition and post-condition checking.
    """

    @staticmethod
    def require(requirement: bool, msg: str):
        """
        Checks that given requirement is true, throws an ValueError if false
        :param requirement: testable requirement
        :param msg: string that is used to build exception message
        """
        if not requirement:
            raise ValueError('requirement failed: ' + msg)

    @staticmethod
    def require_not_null(arg, arg_name: str):
        """
        Checks and throws an ValueError if the given object is null
        :param arg: any object that can be null
        :param arg_name: name of the argument being checked, used in exception message if null
        """
        DbcUtil.require(arg is not None, 'required that ' + arg_name + ' must NOT be None')

    @staticmethod
    def require_string_not_null_or_empty(arg: str, arg_name: str):
        """
        Checks and throws an ValueError if the given string is null or empty
        :param arg: String argument which needs to be checked
        :param arg_name: name of the argument being checked, used in exception message if null or empty
        """
        DbcUtil.require(arg is not None and len(arg) > 0, 'required that ' + arg_name + ' must NOT be null or empty')

    @staticmethod
    def require_array_not_contain_null(array: list, arg_name: str):
        """
        Checks and throws an ValueError if the given array is null or contains null element
        :param array: Array argument which needs to be checked
        :param arg_name: name of the argument being checked, used in exception message
        """
        DbcUtil.require_not_null(array, arg_name)
        for x in array:
            DbcUtil.require(x, 'required that elements of ' + arg_name + ' must NOT be None')
